hermes/crates/cardano-chain-follower/src/metadata/cip509/mod.rs

```diff
@@ -0,0 +1,954 @@
+
+// cspell: words pkix
+use c509_certificate::general_names::general_name::GeneralNameValue;
+use decode_helper::{decode_bytes, decode_map_len, decode_u8};
+use der_parser::{asn1_rs::oid, der::parse_der_sequence, Oid};
+use rbac::{certs::C509Cert, role_data::RoleData};
+
+mod decode_helper;
+mod rbac;
+use x509_cert::{der::Decode as _, ext::pkix::ID_CE_SUBJECT_ALT_NAME};
+mod x509_chunks;
+
+use std::sync::Arc;
+
+use minicbor::{
+    decode::{self},
+    Decode, Decoder,
+};
+use pallas::{
+    codec::{
+        minicbor::{Encode, Encoder},
+        utils::Bytes,
+    },
+    ledger::traverse::MultiEraTx,
+};
+use strum::FromRepr;
+use x509_chunks::X509Chunks;
+
+use super::{
+    raw_aux_data::RawAuxData, DecodedMetadata, DecodedMetadataItem, DecodedMetadataValues,
+    ValidationReport,
+};
+use crate::{
+    utils::{
+        blake2b_128, blake2b_256, compare_key_hash, decode_utf8, extract_cip19_hash,
+        extract_key_hash, zero_out_last_n_bytes,
+    },
+    witness::TxWitness,
+    Network,
+};
+
+/// CIP509 label.
+pub const LABEL: u64 = 509;
+
+/// Context-specific primitive type with tag number 6 (`raw_tag` 134) for
+/// uniform resource identifier (URI) in the subject alternative name extension.
+pub(crate) const URI: u8 = 134;
+
+/// Subject Alternative Name OID
+pub(crate) const SUBJECT_ALT_NAME_OID: Oid = oid!(2.5.29 .17);
+
+/// CIP509 metadatum.
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct Cip509 {
+    /// `UUIDv4` Purpose .
+    pub purpose: [u8; 16], // (bytes .size 16)
+    /// Transaction inputs hash.
+    pub txn_inputs_hash: [u8; 16], // bytes .size 16
+    /// Optional previous transaction ID.
+    pub prv_tx_id: Option<[u8; 32]>, // bytes .size 32
+    /// x509 chunks.
+    pub x509_chunks: X509Chunks, // chunk_type => [ + x509_chunk ]
+    /// Validation signature.
+    pub validation_signature: Vec<u8>, // bytes size (1..64)
+    /// Validation value, not a part of CIP509, justs storing validity of the data.
+    pub validation: Cip509Validation,
+}
+
+/// Validation value for CIP509 metadatum.
+#[allow(clippy::struct_excessive_bools, clippy::module_name_repetitions)]
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct Cip509Validation {
+    /// Boolean value for the validity of the transaction inputs hash.
+    pub valid_txn_inputs_hash: bool,
+    /// Boolean value for the validity of the auxiliary data.
+    pub valid_aux: bool,
+    /// Bytes of precomputed auxiliary data.
+    pub precomputed_aux: Vec<u8>,
+    /// Boolean value for the validity of the public key.
+    pub valid_public_key: bool,
+    /// Boolean value for the validity of the payment key.
+    pub valid_payment_key: bool,
+}
+
+/// Enum of CIP509 metadatum with its associated unsigned integer value.
+#[allow(clippy::module_name_repetitions)]
+#[derive(FromRepr, Debug, PartialEq)]
+#[repr(u8)]
+pub(crate) enum Cip509Int {
```

Comment:
Does not look like as an appropriate name `Cip509Int`, it looks as a type identifier flag not as a integer value.